cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/RawMetricValues.java

```diff
@@ -0,0 +1,351 @@
+    boolean allIndexesValid = _validity.cardinality() - numValidIndexesAdjustment == _counts.length - 1;
+
+    boolean tooManyImputations = numImputations() > maxAllowedImputation;
+
+    return allIndexesValid && !tooManyImputations;
+  }
+
+  public synchronized int numImputations() {
+    int currentIdx = handleWrapping(currentWindowIndex());
+    int numImputationAdjustment = _imputations.get(currentIdx) ? 1 : 0;
+    return _imputations.cardinality() - numImputationAdjustment;
+  }
+
+  public synchronized boolean isValidAtWindowIndex(long windowIndex) {
+    validateIndex(windowIndex);
+    return _validity.get(handleWrapping(windowIndex));
+  }
+
+  public synchronized boolean hasImputationAtWindowIndex(long windowIndex) {
+    validateIndex(windowIndex);
+    return _imputations.get(handleWrapping(windowIndex));
+  }
+
+  public synchronized int sampleCountsAtWindowIndex(long windowIndex) {
+    validateIndex(windowIndex);
+    return _counts[handleWrapping(windowIndex)];
+  }
+
+  public synchronized void resetWindowIndexes(long startingWindowIndex, int numWindowIndexesToReset) {
+    if (inValidRange(startingWindowIndex)
+        || inValidRange(startingWindowIndex + numWindowIndexesToReset - 1)) {
+      throw new IllegalStateException("Should never reset a window index that is in the valid range");
+    }
+    LOG.debug("Resetting window index [{}, {}]", startingWindowIndex,
+              startingWindowIndex + numWindowIndexesToReset - 1);
+    // We are not resetting all the data here. The data will be interpreted to 0 if count is 0.
+    for (long i = startingWindowIndex; i < startingWindowIndex + numWindowIndexesToReset; i++) {
+      int index = handleWrapping(i);
+      _counts[index] = 0;
+      _validity.clear(index);
+      _imputations.clear(index);
+    }
+  }
+
+  public synchronized ValuesAndImputations aggregate(SortedSet<Long> windowIndexes, MetricDef metricDef) {
+    if (_values.isEmpty()) {
+      return ValuesAndImputations.empty(windowIndexes.size(), metricDef);
+    }
+    Map<Integer, MetricValues> aggValues = new HashMap<>();
+    SortedMap<Integer, Imputation> imputations = new TreeMap<>();
+    for (Map.Entry<Integer, float[]> entry : _values.entrySet()) {
+      int metricId = entry.getKey();
+      float[] values = entry.getValue();
+      MetricInfo info = metricDef.metricInfo(metricId);
+
+      MetricValues aggValuesForMetric = new MetricValues(windowIndexes.size());
+      aggValues.put(metricId, aggValuesForMetric);
+
+      int resultIndex = 0;
+      for (long windowIndex : windowIndexes) {
+        validateIndex(windowIndex);
+        int idx = handleWrapping(windowIndex);
+        // Sufficient samples
+        if (_counts[idx] >= _minSamplesPerSnapshotWindow) {
+          aggValuesForMetric.set(resultIndex, getValue(info, idx, values));
+        // Not quite sufficient, but have some available.
+        } else if (_counts[idx] >= halfMinRequiredSamples()) {
+          if (info.id() == 0) {
```

Comment:
This was to avoid putting the imputations multiple times if there are more than one metric type to aggregate. But it should probably be putIfAbsent()...